core/src/main/java/com/linecorp/armeria/server/throttling/FileBasedDynamicThrottlingStrategy.java

```diff
@@ -0,0 +1,157 @@
+ * under the License.
+ */
+package com.linecorp.armeria.server.throttling;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+import java.io.IOException;
+import java.nio.file.ClosedWatchServiceException;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ThreadLocalRandom;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Ascii;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.primitives.Floats;
+import com.sun.nio.file.SensitivityWatchEventModifier;
+
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.RpcRequest;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+/**
+ * A {@link ThrottlingStrategy} to read and maintain a map from service API name to {@link ThrottlingStrategy}.
+ * <p>
+ * The {@link FileBasedDynamicThrottlingStrategy} will monitor a json mapping file for updates, whose format is
+ * keys for every service API name with values is a string representation of {@link ThrottlingStrategy}.
+ */
+public class FileBasedDynamicThrottlingStrategy<T extends Request> extends ThrottlingStrategy<T> {
+    private static final Logger logger = LoggerFactory.getLogger(FileBasedDynamicThrottlingStrategy.class);
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    private static final Splitter COLON_SPLITTER = Splitter.on(':').trimResults();
+
+    private final WatchService watchService;
+    private final Path settingFile;
+    private final Thread watchThread;
+    private volatile Map<String, ThrottlingStrategy<T>> strategies = ImmutableMap.of();
+
+    /**
+     * Creates a new instance.
+     */
+    public FileBasedDynamicThrottlingStrategy(String settingFile) {
+        this.settingFile = FileSystems.getDefault().getPath(settingFile);
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+            this.settingFile.getParent().register(watchService,
+                                                  new WatchEvent.Kind[] {
+                                                          StandardWatchEventKinds.ENTRY_CREATE,
+                                                          StandardWatchEventKinds.ENTRY_MODIFY
+                                                  },
+                                                  SensitivityWatchEventModifier.HIGH);
```

Comment:
It looks like this will poll the file system every 2 seconds. Maybe this excessive? or should be configurable at least?